google/cloud/bigquerycontrol/v2/samples/job_query_samples.cc

```
@@ -0,0 +1,160 @@
+
+  //! [execute-query-job]
+  namespace bigquery_v2_proto = google::cloud::bigquery::v2;
+  [](std::string const& billing_project, std::string const& query,
+     int retry_delay) {
+    auto client = google::cloud::bigquerycontrol_v2::JobServiceClient(
+        google::cloud::bigquerycontrol_v2::MakeJobServiceConnectionRest());
+
+    bigquery_v2_proto::JobConfigurationQuery query_config;
+    // Using GoogleSQL allows instance and dataset specification in the FROM
+    // clause as part of the table name.
+    query_config.mutable_use_legacy_sql()->set_value(false);
+    query_config.set_query(query);
+
+    bigquery_v2_proto::JobConfiguration config;
+    *config.mutable_query() = query_config;
+    config.mutable_labels()->emplace("type", "sample");
+    bigquery_v2_proto::Job job;
+    *job.mutable_configuration() = config;
+    bigquery_v2_proto::InsertJobRequest job_request;
+    job_request.set_project_id(billing_project);
+    *job_request.mutable_job() = job;
+
+    // Submit the Query Job.
+    auto insert_result = client.InsertJob(job_request);
+    if (!insert_result) throw insert_result.status();
+
+    auto job_id = insert_result->job_reference().job_id();
+    // If not immediately finished, poll job status until DONE.
+    if (insert_result->status().state() != "DONE") {
+      bigquery_v2_proto::GetJobRequest get_request;
+      get_request.set_project_id(billing_project);
+      get_request.set_job_id(job_id);
+
+      bool job_complete = false;
+      std::vector<int> retries{retry_delay, retry_delay, retry_delay,
+                               retry_delay, retry_delay};
+      for (auto delay : retries) {
+        auto get_result = client.GetJob(get_request);
+        if (get_result->status().state() == "DONE") {
+          job_complete = true;
+          break;
+        }
+        std::this_thread::sleep_for(std::chrono::seconds(delay));
+      }
+      if (!job_complete) throw "Query execution timed out after 5 retries.";
```

Better to throw `std::runtime_error()` I think strings will print as "unknown error" without any details.